File: 100-Days-Series/Day 73/ValidParentheses.js
class ValidParentheses{
    static isValidParentheses(str){
        const stack=[];// Use an array as a stack
        for(let ch of str){
            if(ch=='(' || ch=='[' || ch=='{'){
                stack.push(ch);        
            }else if(ch==')' || ch==']' || ch=='}'){
                if(stack.length===0 || !this.isMatchingPair(stack[stack.length-1],ch))
                    return false;
                stack.pop();
            }
        }
        return stack.length===0;
    }
    static isMatchingPair(open,close){
        return (open==='(' && close===')') ||
            (open==='[' && close===']') ||
            (open==='{' && close==='}');
    }
}
// Let's Start Here
const str="()[]{}";
const res=ValidParentheses.isValidParentheses(str);
if(res){
    console.log("Valid Parentheses");
}else{
    console.log("Invalid Parentheses");
}

File: 100-Days-Series/Day 73/ValidParentheses.java
class ValidParentheses{
    public static boolean isValidParentheses(String str){
        java.util.Stack<Character> stack=new java.util.Stack<>();
        for(char ch:str.toCharArray()){
            if(ch=='(' || ch=='[' || ch=='{'){
                stack.push(ch);
            }else if(ch==')' || ch==']' || ch=='}'){
                if(stack.isEmpty() || !isMatchingPair(stack.peek(), ch)){
                    return false;
                }
                stack.pop();
            }
        }
        return stack.isEmpty();
    }
    public static boolean isMatchingPair(char open,char close){
        return (open=='(' && close==')')||
                (open=='[' && close==']')||
                (open=='{' && close=='}');
    }
    // Main Method
    public static void main(String [] args){
        // Let's Start Here
        String str="()[]{}";
        boolean res=isValidParentheses(str);
        if(res){
            System.out.println("Valid Parentheses");
        }else{
            System.out.println("Invalid Parentheses");
        }
    }
}